test/test_python_dispatch.py

```diff
@@ -352,6 +363,179 @@ def sqsum(a: SymInt, b: SymInt):
+        with self.assertRaisesRegex(NotImplementedError, 'no Tensor inputs'):
+            CustomOp.define('a::g() -> Tensor')
+        with self.assertRaisesRegex(NotImplementedError, 'no Tensor inputs'):
+            CustomOp.define('a::h(int[] shape) -> Tensor')
+        with self.assertRaisesRegex(NotImplementedError, 'no outputs'):
+            CustomOp.define('a::i(Tensor x) -> ()')
+
+    def test_supported_schemas(self):
+        # All of these should already be tested by PyTorch codegen
+        # (we share the same mechanism), but here's a sanity check.
+        schemas = [
+            'a::b(Tensor x) -> Tensor',
+            'a::b.with_overload(Tensor x) -> Tensor y',
+            'a::b(Tensor[] x) -> Tensor y',
+            'a::b(Tensor x) -> (Tensor, Tensor)',
+            'a::b(Tensor x) -> (Tensor y, Tensor z)',
+            'a::b(Tensor x) -> (Tensor y, Tensor z)',
+            'a::b(Tensor x, Tensor w) -> (Tensor y, Tensor z)',
+            'a::b.with_overload(Tensor x, Tensor w) -> (Tensor, Tensor)',
+            'a::b(Tensor x, Tensor w) -> Tensor',
+            'a::b(Tensor? x, Tensor w) -> Tensor',
+            'a::b(Tensor? x, Tensor[] w) -> Tensor',
+            'a::b(Tensor x, int[] w) -> Tensor',
+            'a::b(Tensor x, SymInt[] w) -> Tensor',
+            'a::b(Tensor x, Scalar w) -> Tensor',
+            'a::b(Tensor x, float w) -> Tensor',
+            'a::b(Tensor x, float? w) -> Tensor',
+            'a::b(Tensor x, bool[] w) -> Tensor',
+        ]
+
+        for schema in schemas:
+            CustomOp.define(schema)
+
+    def test_reserved_ns(self):
+        from torch._custom_op import RESERVED_NS
+
+        for ns in RESERVED_NS:
+            with self.assertRaisesRegex(ValueError, 'is a reserved namespace'):
+                CustomOp.define(f'{ns}::a(Tensor x) -> Tensor')
+
+    def test_private_ctor(self):
+        with self.assertRaisesRegex(RuntimeError, 'CustomOp constructor is private'):
+            CustomOp(None, None, None)
+
+    def test_lifetime_is_tied_to_object(self):
+        foo = CustomOp.define('custom::foo(Tensor x) -> Tensor')
+
+        # If there is more than one reference then we've got a problem - something
+        # will keep the object alive.
+        # NB: sys.getrefcount(foo) reports one more than the number (since it creates a ref)
+        self.assertEqual(sys.getrefcount(foo), 2)
+
+        # We can't define an op multiple times,
+        with self.assertRaisesRegex(RuntimeError, 'multiple times'):
+            bar = CustomOp.define('custom::foo(Tensor x) -> Tensor')
+
+        # Unless we delete the original op.
+        del foo
+        baz = CustomOp.define('custom::foo(Tensor x) -> Tensor')
+
+    def test_autograd_notimplemented(self):
+        foo = CustomOp.define('custom::foo(Tensor x) -> Tensor')
+        x = torch.randn(3, requires_grad=True)
+        with self.assertRaisesRegex(RuntimeError, 'Autograd has not been implemented'):
+            foo(x)
```

Comment:
The C++ "not implemented" autograd kernel is a bit lazier - you can run the forward without error, but install a "NotImplemented" node in the autograd graph - so it will only yell at you if you later try to backprop through your op. Do you think we should do the same for these custom ops?

Reply:
Yes, we should install the C++ "not implemented" autograd kernel. My plan was to handle this in a follow-up because this PR is getting long by itself and it's not trivial to use the C++ "not implemented" autograd kernel (because there are two of them, one for Autograd and one for ADInplaceAndView) and I need to figure out how to call them from Python